iot_packet_verifier/src/pending.rs

```diff
@@ -0,0 +1,464 @@
+use async_trait::async_trait;
+use chrono::{DateTime, Duration, Utc};
+use helium_crypto::PublicKeyBinary;
+use solana::SolanaNetwork;
+use solana_sdk::signature::Signature;
+use sqlx::{postgres::PgRow, FromRow, PgPool, Postgres, Row, Transaction};
+use std::{collections::HashMap, sync::Arc};
+use tokio::sync::Mutex;
+
+/// To avoid excessive burn transaction (which cost us money), we institute a minimum
+/// amount of Data Credits accounted for before we burn from a payer:
+const BURN_THRESHOLD: i64 = 10_000;
+
+#[async_trait]
+pub trait AddPendingBurn {
+    async fn add_burned_amount(
+        &mut self,
+        payer: &PublicKeyBinary,
+        amount: u64,
+    ) -> Result<(), sqlx::Error>;
+}
+
+#[async_trait]
+pub trait PendingTables {
+    type Transaction<'a>: PendingTablesTransaction<'a> + Send + Sync
+    where
+        Self: 'a;
+
+    async fn fetch_next_burn(&self) -> Result<Option<Burn>, sqlx::Error>;
+
+    async fn fetch_all_pending_burns(&self) -> Result<Vec<Burn>, sqlx::Error>;
+
+    async fn fetch_all_pending_txns(&self) -> Result<Vec<PendingTxn>, sqlx::Error>;
+
+    async fn submit_txn(
+        &self,
+        payer: &PublicKeyBinary,
+        amount: u64,
+        signature: &Signature,
+    ) -> Result<(), sqlx::Error>;
+
+    async fn begin<'a>(&'a self) -> Result<Self::Transaction<'a>, sqlx::Error>;
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum ConfirmPendingError<S> {
+    #[error("Sqlx error: {0}")]
+    SqlxError(#[from] sqlx::Error),
+    #[error("Chrono error: {0}")]
+    ChronoError(#[from] chrono::OutOfRangeError),
+    #[error("Solana error: {0}")]
+    SolanaError(S),
+}
+
+pub async fn confirm_pending_txns<S>(
+    pending_tables: &impl PendingTables,
+    solana: &S,
+) -> Result<(), ConfirmPendingError<S::Error>>
+where
+    S: SolanaNetwork,
+{
+    // Fetch all pending transactions and confirm them
+    let pending = pending_tables.fetch_all_pending_txns().await?;
+    for pending in pending {
```

Comment:
Two reasons. The first is for testing, so that I can create a unit test that completes instantly.
The second is that it allows us to start the iot packet verifier instantly when there are no outstanding transactions

Reply:
what about `if pending.len() > 0 { sleep_for_1_min_only_once }`? that should address the second point and if you have `fetch_all_pending_txns` order the results by `time_of_submission` then you can just grab the first or last (depending on how you want to order the results) and only sleep for 1 min iff the list is non-empty and the most recent submission was less than 1 min ago